File: python/src/graph/shortest_path.py
import sys

from python.src.util.stack import Stack


class ShortestPath:

    def __init__(self, graph, s):
        self.edge_to = [None] * graph.num_vertices()
        self.dist_to = [0.0] * graph.num_vertices()

        # TODO: complete this function

    def distance_to(self, v):
        return self.dist_to[v]

    def has_path_to(self, v):
        return self.dist_to[v] < sys.maxsize

    def get_path_to(self, v):
        if not self.has_path_to(v):
            return None

        stack = Stack()
        edge = self.edge_to[v]
        while edge is not None:
            stack.push(edge)
            edge = self.edge_to[edge.dest()]

        return stack

File: java/src/edu/berkeley/cs/graph/ShortestPath.java
package edu.berkeley.cs.graph;

import edu.berkeley.cs.util.Stack;

public class ShortestPath {
  private Edge[] edgeTo;
  private float[] distTo;

  /**
   * @param graph the graph to search through
   * @param s the source vertex where the search will begin
   */
  public ShortestPath(Graph graph, int s) {
    // TODO: complete this function
  }

  /** @return the distance from the source vertex to vertex v */
  public float distTo(int v) {
    return distTo[v];
  }

  /** @return true if the source vertex has a path to vertex v */
  public boolean hasPathTo(int v) {
    return distTo[v] < Float.MAX_VALUE;
  }

  /** @return the path from the source vertex to vertex v */
  public Iterable<Edge> pathTo(int v) {
    if (!hasPathTo(v)) {
      return null;
    }

    Stack<Edge> stack = new Stack<>();

    Edge edge = edgeTo[v];
    while (edge != null) {
      stack.push(edge);
      edge = edgeTo[edge.to()];
    }

    return stack;
  }
}
